### ISSUE TITLE
Request Uri problem in controller tests

### ISSUE THREAD

**Issue by SpiGAndromeda** [Shardj/zf1-future#52]
Hi,

I am currently writing some smoke tests for an old ZF1 application. I encountered a problem with the REQUEST URI generation of ZF. If dispatching an action via `Zend_Test_PHPUnit_ControllerTestCase`, the URI `$_SERVER['REQUEST_URI']` generation is done automatically, which leads to some problems.

The dispatch looks like this:
```php
$url = '/' . $category_slug . '/'. $product_slug;
$this->dispatch($url);
```
The action is correctly dispatched. But the inspection of `$this` in the .phtml temaplate shows:
`message="Invalid URI supplied"`.

I digged into ZF1 and found the reason in the `library/Zend/Uri/Http.php` file:
```php
public function validateHost($host = null)
    {
        if ($host === null) {
            $host = $this->_host;
        }

        // If the host is empty, then it is considered invalid
        if (strlen($host) === 0) {
            return false;
        }

        // Check the host against the allowed values; delegated to Zend_Filter.
        $validate = new Zend_Validate_Hostname(Zend_Validate_Hostname::ALLOW_ALL);

        return $validate->isValid($host);
    }
```
The dispatch method sets the Request URI by itself. I already tried setting it manually, but it's overwritten:
```php
// same without scheme
$this->getRequest()->setRequestUri('http://host.domain:port' . $url);
```
The URI is evaluated, and the empty check triggers.
Passing a complete URI (with host) to dispatch tirggers also: `message="Invalid controller specified"`.

So basically if you evaluate the `$_SERVER['REQUEST_URI']` variable in the template (I know, I don't like it either, but it's there, currently), you will always throw in a PHPUnit test.

I spontaniusly see three ways of solving this problem (if it is a problem at all and I am not getting something terribly wrong):
*  adding a PHPUnit flag like the one in `Zend_Session` to bypass the the URI valid check (but that would dig pretty deep in ZF domain object validation)
*  adding a condition to the dispatch method which considers an already set RequestUri:
```php
// lines 178 - 181
$request    = $this->getRequest();
// a valid URI would be set before calling dispatch
if (null !== $url && null === $request->getRequestUri()) {
    $request->setRequestUri($url);
}
```
*  adding an optional argument to the dispatch method to supply it with manually set `$_SERVER` parameters. Default is `null` (which is checked before setting something) to avoid BC breaks

best regards,
spigandromeda

P.S. it also took me a while to find a working package combination for using PHPUnit at all. Maybe you want to add this somewhere in the documentation:
```json
"require-dev": {
    "phpunit/dbunit": "2.0.3",
    "phpunit/php-invoker": "1.1.4",
     "phpunit/phpunit": "5.3.5"
}
```
That seems to be the most up-to-date version of PHPUnit working with ZF1 (ZF1 tests are unsing DBUnit tests, which is deprecated since two years or something and this resitrcts the PHPUnit version).

P.P.S I really appreciate your work! I bet there are tons of old projects still running on ZF1 which would starve for the necessary updates to use PHP7 and composer. <3

**Comment by kaktusas2598** [comment]
Could you share full source code of unit test class you are having problems with?

**Comment by Shardj** [comment]
Sorry but I'm a bit confused by what you're doing, it's not something I've ever attempted. Could you paste your test class and any other files needed so that I can go and reproduce this issue?

I don't know about phpunit/dbunit or /php-invoker. But I'm using phpunit version 6.5.14 without issue.

**Comment by SpiGAndromeda** [comment]
@Shardj sure

It's just a Smoke test. So it should draw all products from the databse and then dispatch the route for everyone. Asserts will be added when the dispatch is working without errors.

```php
<?php

declare(strict_types=1);

final class ProductControllerTest extends Zend_Test_PHPUnit_ControllerTestCase
{
    /**
     * @throws Zend_Application_Exception
     */
    public function setUp(
    ): void
    {
        // disables some
        Zend_Session::$_unitTestEnabled = true;

        $this->bootstrap = new Zend_Application('testing',
            APPLICATION_PATH . '/configs/application.ini');
        parent::setUp();
    }

    public function testIndexAction(
    ): void
    {
        $productClass = new products();
        $categoryClass = new categories();

        $products = $productClass->fetchAll(['product_active = ?' => 1], null, 20);
        foreach ($products as $product) {
            /** @var Vorlagen_Db_Table_Row_Product $product */

            /** @var string $id_category */
            $id_category = $product->id_category;

            /** @var Vorlagen_Db_Table_Row_Category $category */
            $category = $categoryClass->load((int) $id_category);

            /** @var string $product_slug */
            $product_slug = $product->product_slug;
            /** @var string $category_slug */
            $category_slug = $category->category_slug;

            $url = '/' . $category_slug . '/'. $product_slug;
            $this->dispatch($url);
        }
    }
}
```

I tested it with PHPUnit 6.5.14: `Fatal error: Uncaught Error: Class 'PHPUnit_Framework_TestCase' not found in /var/www/html/vendor/shardj/zf1-future/library/Zend/Test/PHPUnit/ControllerTestCase.php on line 47`

The ZF1 Test classes seems to rely on the Zend like syntax for class naming and folder structure which was changed in PHPUnit by v6 (I guess).

**Comment by Shardj** [comment]
I'll have a look at reproducing this tomorrow if I have time. In the meantime, you can fix that error you got with 6.5.14 with a bootstrap file:

```
<?php

define('TEST_EXECUTION', true);

require_once dirname(__FILE__) . '/../..' . '/app-loader.php';
class PHPUnit_Framework_TestCase extends \PHPUnit\Framework\TestCase {}
```

**Comment by Shardj** [comment]
Closing for inactivity
